File: web_api/schemas/product_sync.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ProductSyncCreateRequest(BaseModel):
    mode: str = "full"
    operator: str = "local_admin"
    limit: int | None = Field(default=None, ge=1, le=500)


class ProductSyncJobItemResponse(BaseModel):
    id: str
    job_id: str
    shop_id: str
    goods_id: str | None = None
    goods_name: str | None = None
    status: str
    error_summary: str | None = None
    created_at: str
    updated_at: str


class ProductSyncJobResponse(BaseModel):
    id: str
    shop_id: str
    internal_shop_id: int | None = None
    status: str
    mode: str
    total_count: int = 0
    succeeded_count: int = 0
    failed_count: int = 0
    skipped_count: int = 0
    coverage_json: dict[str, Any] | None = None
    error_summary: str | None = None
    created_by: str | None = None
    created_at: str
    started_at: str | None = None
    finished_at: str | None = None
    updated_at: str


class ProductSyncJobDetailResponse(ProductSyncJobResponse):
    items: list[ProductSyncJobItemResponse] = Field(default_factory=list)


class ProductSyncCoverageResponse(BaseModel):
    shop_id: str
    total: int = 0
    has_price: int = 0
    has_specs: int = 0
    has_usage: int = 0
    has_ingredients: int = 0
    has_shelf_life: int = 0
    has_warnings: int = 0
    has_manual_notes: int = 0
    last_sync_at: str | None = None
    warning: str | None = None

File: web_frontend/src/api/productSync.ts
import { apiGet, apiPost } from "./client";

export type ProductSyncJobItem = {
  id: string;
  job_id: string;
  shop_id: string;
  goods_id?: string | null;
  goods_name?: string | null;
  status: string;
  error_summary?: string | null;
  created_at: string;
  updated_at: string;
};

export type ProductSyncJob = {
  id: string;
  shop_id: string;
  internal_shop_id?: number | null;
  status: string;
  mode: string;
  total_count: number;
  succeeded_count: number;
  failed_count: number;
  skipped_count: number;
  coverage_json?: Record<string, unknown> | null;
  error_summary?: string | null;
  created_by?: string | null;
  created_at: string;
  started_at?: string | null;
  finished_at?: string | null;
  updated_at: string;
  items?: ProductSyncJobItem[];
};

export type ProductSyncCoverage = {
  shop_id: string;
  total: number;
  has_price: number;
  has_specs: number;
  has_usage: number;
  has_ingredients: number;
  has_shelf_life: number;
  has_warnings: number;
  has_manual_notes: number;
  last_sync_at?: string | null;
  warning?: string | null;
};

export function createProductSyncJob(shopId: string, limit = 10) {
  return apiPost<ProductSyncJob>(`/api/shops/${encodeURIComponent(shopId)}/product-sync/jobs`, {
    mode: "full",
    operator: "local_admin",
    limit
  });
}

export function listProductSyncJobs(shopId: string, limit = 20) {
  return apiGet<ProductSyncJob[]>(`/api/shops/${encodeURIComponent(shopId)}/product-sync/jobs?limit=${limit}`);
}

export function getProductSyncJob(shopId: string, jobId: string) {
  return apiGet<ProductSyncJob>(`/api/shops/${encodeURIComponent(shopId)}/product-sync/jobs/${encodeURIComponent(jobId)}`);
}

export function retryProductSyncJob(shopId: string, jobId: string) {
  return apiPost<ProductSyncJob>(`/api/shops/${encodeURIComponent(shopId)}/product-sync/jobs/${encodeURIComponent(jobId)}/retry`, {});
}

export function getProductSyncCoverage(shopId: string) {
  return apiGet<ProductSyncCoverage>(`/api/shops/${encodeURIComponent(shopId)}/product-sync/coverage`);
}
